### PR TITLE
feat: add items api to context

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
## Description

While developing the [twisted integration](https://github.com/Datadog/dd-trace-py/pull/1669)
the need arose for the integration to add and retrieve items from the context
in order to trace `Deferred` objects. This API provides a simple way to do
this.

## Checklist
- [x] Entry added to `CHANGELOG.md`.
- [x] Tests provided; and/or
- [x] ~~Description of manual testing performed and explanation is included in the code and/or PR.~~
- [x] ~~Library documentation is updated.~~ (done automatically)
- [x] ~~[Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~~ N/A

### CODE REVIEW

File: ddtrace/context.py [hunk 1]
```diff
@@ -69,6 +71,30 @@ def sampling_priority(self, value):
         with self._lock:
             self._sampling_priority = value
 
+    def set_ctx_item(self, key, val):
```
Comment (jd) [inline ddtrace/context.py:74]: Any reason not to use `__getitem__` and `__setitem__`?

**Review by jd — COMMENTED** [review]
What's the final goal? :)

**Review by jd — COMMENTED** [review]
Honestly, I'm not a fan of having a kind of dumping ground in a context like that.
This is going to end up being a mess and might even lead to memory leaks.

In terms of performance, using the same lock as the rest of the structure where as the `_ctx` dict does not need to use the same one, is also a bad point.

File: ddtrace/context.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply ddtrace/context.py:74]: Yeah I debated going this approach. I guess it makes sense.

**Comment by Kyle-Verhoog** [comment]
The alternative is that we have each integration that wants to store something that depends on context itself or add the fields explicitly to the context (I had originally done this but it seems messy). I agree that it could become messy if abused but the cases in which is needed are pretty uncommon and should be well reasoned about.

I think both OpenTelemetry and OpenTracing have this idea of context items so there is another precedent for it.

**Comment by brettlangdon** [comment]
Not sure I fully get what is going on here, basically we need something similar to Flask's request context for libraries that don't offer that as part of it's API?

**Comment by Kyle-Verhoog** [comment]
The motivation is so that integrations (or users) can add things to the context so they can be retrieved at some point later. My use-case for Twisted is to enable `Deferred` objects to be [traced in certain contexts](https://github.com/DataDog/dd-trace-py/pull/1669/commits/f91184ada1c41ddfcb1a490a7c0452b1b42b7fb1#diff-8d9adb60bb445684775af48bd4332529R104-R106).

```python
    ctx = pin.tracer.get_call_context()
    with ctx.override_ctx_item("trace_deferreds", True):
        with ctx.override_ctx_item("deferred_name", "runQuery"):
            return func(*args, **kwargs)
```

```python
    ddctx = ddtrace.tracer.get_call_context()
    if ddctx.get_ctx_item("trace_deferreds", default=False):
        name = ddctx.get_ctx_item("deferred_name")
        span = pin.tracer.trace(name)
```

**Pull request closed without merging**